### ISSUE TITLE
Incorrect error reporting on global variables

### ISSUE THREAD

**Issue by IngwiePhoenix** [vlang/v#7662]
**V version:** 
**OS:** Windows

**What did you do?**
Compiled an example piece i made.

**What did you expect to see?**
An error indicating the illegality of global variables
 
**What did you see instead?**
```
global.v:3:4: error: unexpected name `main`, expecting `(`
    1 | mynum := 10
    2 |
    3 | fn main() {
      |    ~~~~
    4 | println(mynum)
    5 | }
```
(1-5 is actually the entire code.)

**Comment by danieldaeschle** [comment]
when you declare a `mynum := 10` V treats your file as a script. That means a main function is invalid.

**Comment by danieldaeschle** [comment]
the error should be more detailed.

**Comment by danieldaeschle** [comment]
if you declare mynum after main, you get your expected error

**Comment by JalonSolov** [comment]
Even if that particular case said something like `fn main is not valid in a script` or something, instead of just `unexpected main`, it would give more of a clue to what is wrong.

**Comment by danieldaeschle** [comment]
Or maybe main function should work and be just a normal function which can be called with `main()`

**Comment by JalonSolov** [comment]
Please no.  That would make it different from just about every other language, for no good reason.

**Comment by spytheman** [comment]
There is no global level program.
The way V works, is that if there are top level statements, it treats them as if they were inside an implicit `fn main() {}`.
Then the `fn main() {}` that is defined by the user after the top level assignment, becomes a function statement, inside another function => the error.

**Comment by danieldaeschle** [comment]
> Please no. That would make it different from just about every other language, for no good reason.

yeah, i found out that no function declarations work in script mode. the error needs to be different.



> There is no global level program.
> The way V works, is that if there are top level statements, it treats them as if they were inside an implicit `fn main() {}`.
> Then the `fn main() {}` that is defined by the user after the top level assignment, becomes a function statement, inside another function => the error.
> 
> TLDR: V uses a heuristic, and that heuristic fails. In my opinion all programs should use main and not top level assignments, even .vsh scripts. That will remove any guessing and confusion. On the negative side, simple programs like just `println('hi')` or `2+2` would no longer compile, they would need to be wrapped in `fn main() {` and `}`.

In my opinion .vsh should not need fn main() but it should allow function and type declaration. But `.v` should have a `main()` every time.
